desc/grid.py

```diff
@@ -447,31 +450,62 @@ def _create_nodes(
             if self.sym:
                 t += t[1] / 2
             dt = 2 * np.pi / t.size * np.ones_like(t)
+            # if endpoint:
+            #     FIXME: dt already has the correct weights
+            #         _scale_weights() function is about to mess it up
+            #         can test with endpoint = True, ntheta = 3
+
         else:
             t = np.atleast_1d(theta)
             dt = np.zeros_like(t)
             if t.size > 1:
-                dt[0] = (t[0] + t[1]) / 2
-                dt[1:-1] = (t[2:] - t[:-2]) / 2
-                dt[-1] = 2 * np.pi - (t[-2] + t[-1]) / 2
+                # choose dt to be half the cyclic distance of the surrounding two nodes
+                SUP = 2 * np.pi  # supremum
+                dt[0] = t[1] + (SUP - (t[-1] % SUP)) % SUP
+                dt[1:-1] = t[2:] - t[:-2]
+                dt[-1] = t[0] + (SUP - (t[-2] % SUP)) % SUP
+                dt /= 2
+                if t.size == 2:
+                    dt[-1] = dt[0]
+                # if endpoint and dt[0] == (dt[-1] % SUP):
+                #     FIXME: dt already has the correct spacing
+                #         _scale_weights() function is about to mess it up
+                #         can test with endpoint = True, ntheta = 3
             else:
                 dt = np.array([2 * np.pi])
 
         # zeta
+        # note: dz spacing should not depend on NFP
+        # spacing corresponds to a node's weight in an integral --
+        # such as integral = sum(dz * data["B"]) -- not the node's coordinates
         if self.N is not None:
             zeta = 2 * self.N + 1
         else:
             self._N = len(np.atleast_1d(zeta))
         if np.isscalar(zeta) and (int(zeta) == zeta) and zeta > 0:
             z = np.linspace(0, 2 * np.pi / self.NFP, int(zeta), endpoint=endpoint)
             dz = 2 * np.pi / z.size * np.ones_like(z)
+            # if endpoint:
+            #     FIXME: dz already has the correct spacing
+            #         _scale_weights function() is about to mess it up
+            #         can test with endpoint = True, nzeta = 3
         else:
             z = np.atleast_1d(zeta)
             dz = np.zeros_like(z)
             if z.size > 1:
-                dz[0] = (z[0] + z[1]) / 2
-                dz[1:-1] = (z[2:] - z[:-2]) / 2
-                dz[-1] = 2 * np.pi - (z[-2] + z[-1]) / 2
+                # choose dz to be half the cyclic distance of the surrounding two nodes
+                SUP = 2 * np.pi / self.NFP  # supremum
+                dz[0] = z[1] + (SUP - (z[-1] % SUP)) % SUP
+                dz[1:-1] = z[2:] - z[:-2]
+                dz[-1] = z[0] + (SUP - (z[-2] % SUP)) % SUP
+                dz /= 2
+                dz *= self.NFP
+                if z.size == 2:
+                    dz[-1] = dz[0]
+                # if endpoint and dz[0] == (dz[-1] % (SUP * self.NFP)):
+                #     FIXME: dz already has the correct spacing
```

Comment:
Is this something we need to fix now? Do duplicate nodes not get weighted correctly?

Reply:
I put that comment there to remind myself or others to eventually fix it.

Our `_scale_weights` does not reduce duplicate node weight correctly - here or on `master`. The notebook in my above comment can help show what follows.

A simple example of this is `nrho=1, ntheta=5, nzeta=1, endpoint=True`. The correct theta spacing for this `[0.78539816 1.57079633 1.57079633 1.57079633 0.78539816]` with drho = 1 always, dzeta = 2pi always. And at this point in the code that is what the spacing is (also both here and on `master`).

`_scale_weights` then detects the duplicate nodes at 0 and 2pi and scales down the node volume drho * dtheta * dzeta. So the end result of the grid spacing is:
```
[[0.82982653 0.65174424 5.21395388]
 [1.04551592 1.64229256 6.56917025]
 [1.04551592 1.64229256 6.56917025]
 [1.04551592 1.64229256 6.56917025]
 [0.82982653 0.65174424 5.21395388]]
```
The reason I think this is wrong is because a rho surface integral uses the dtheta * dzeta portion of the node volume, ignoring drho. Similar for other surface labels. If we wanted the correct rho surface integral we should leave the drho unchanged and scale up /down dtheta * dzeta.

Reply:
Ahhh I see. Can you make a new issue to keep track of that? Otherwise this PR looks good to go.